File: gems/decomplex/lib/decomplex/oversized_predicate.rb
# frozen_string_literal: true

require_relative "syntax"

module Decomplex
  # Flags boolean predicates with too many independent condition atoms.
  #
  # The intent is not "shorter code"; it is to surface places where a
  # long inline predicate probably wants an existing domain helper or a
  # newly named predicate. Nested parentheses still count because Ruby's
  # AST preserves the same AND/OR tree either way.
  class OversizedPredicate
    LIMIT = 3

    def self.scan(files, limit: LIMIT)
      findings = []
      files.each do |file|
        document = Syntax.parse(file, parser: "tree_sitter")
        new(file, limit).tap do |scanner|
          scanner.collect(document)
          findings.concat(scanner.findings)
        end
      end
      Result.new(findings)
    end

    Result = Struct.new(:findings)

    attr_reader :findings

    def initialize(file, limit)
      @file = file
      @limit = limit
      @findings = []
    end

    def collect(document)
      document.decision_sites.each { |site| record_predicate(site) }
    end

    private

    def record_predicate(site)
      return if predicate_helper?(site.function)

      atoms = condition_atoms(site.predicate)
      return unless atoms.size > @limit

      defn = site.function || "<top>"
      at = "#{@file}:#{defn}:#{site.line}"
      @findings << {
        at: at,
        count: atoms.size,
        predicate: site.predicate,
        atoms: atoms,
        spans: { at => site.enclosing_span || site.span },
      }
    end

    def condition_atoms(predicate)
      predicate.to_s
               .split(/\s*(?:&&|\|\||\band\b|\bor\b)\s*/)
               .map { |atom| atom.gsub(/[()]/, "").strip }
               .reject(&:empty?)
    end

    def predicate_helper?(name)
      name.to_s.end_with?("?")
    end
  end
end

File: gems/decomplex/rust/src/decomplex/detectors/oversized_predicate.rs
use crate::decomplex::ast::Span;
use crate::decomplex::syntax::{self, Document, Language};
use anyhow::Result;
use serde::Serialize;
use std::collections::BTreeMap;
use std::path::PathBuf;

#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
pub struct OversizedPredicateRow {
    pub at: String,
    pub count: usize,
    pub predicate: String,
    pub atoms: Vec<String>,
    pub spans: BTreeMap<String, Span>,
}

#[derive(Clone, Debug, Serialize)]
pub struct ResultReport {
    pub findings: Vec<OversizedPredicateRow>,
}

const LIMIT: usize = 3;
pub fn scan_files(files: &[PathBuf], language: Language) -> Result<ResultReport> {
    let documents = syntax::parse_files(files, language)?;
    Ok(scan_documents(&documents))
}

pub fn scan_documents(documents: &[Document]) -> ResultReport {
    let mut findings = Vec::new();
    for document in documents {
        let scanner = OversizedPredicate::new(LIMIT);
        for site in &document.decision_sites {
            if let Some(finding) = scanner.finding_for_site(site) {
                findings.push(finding);
            }
        }
    }
    ResultReport { findings }
}

struct OversizedPredicate {
    limit: usize,
}

impl OversizedPredicate {
    fn new(limit: usize) -> Self {
        Self { limit }
    }

    fn finding_for_site(
        &self,
        site: &crate::decomplex::syntax::DecisionSite,
    ) -> Option<OversizedPredicateRow> {
        if self.predicate_helper(&site.function) {
            return None;
        }
        let atoms_text = self.condition_atoms(&site.predicate);
        if atoms_text.len() <= self.limit {
            return None;
        }

        let at = format!("{}:{}:{}", site.file, site.function, site.line);
        let mut spans = BTreeMap::new();
        spans.insert(at.clone(), site.enclosing_span);

        Some(OversizedPredicateRow {
            at,
            count: atoms_text.len(),
            predicate: site.predicate.clone(),
            atoms: atoms_text,
            spans,
        })
    }

    fn condition_atoms(&self, predicate: &str) -> Vec<String> {
        predicate
            .split("&&")
            .flat_map(|part| part.split("||"))
            .flat_map(|part| part.split(" and "))
            .flat_map(|part| part.split(" or "))
            .map(|atom| atom.replace(['(', ')'], "").trim().to_string())
            .filter(|atom| !atom.is_empty())
            .collect()
    }

    fn predicate_helper(&self, name: &str) -> bool {
        name.ends_with('?')
    }
}
